exonum-node/src/messages/types.rs

```diff
@@ -0,0 +1,693 @@
+// Copyright 2020 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bit_vec::BitVec;
+use chrono::{DateTime, Utc};
+use exonum::{
+    blockchain::Block,
+    crypto::{Hash, PublicKey},
+    helpers::{Height, Round, ValidatorId},
+    impl_exonum_msg_try_from_signed,
+    merkledb::{BinaryValue, HashTag},
+    messages::{AnyTx, Precommit, SignedMessage},
```

Comment:
A bit of reflections: In theory, `Precommit` is the biggest thing that currently ties us to the BFT consensus (other one is storing `ConsensusConfig` in core schema, but it's much easier to make abstract).
If we'll be able to replace it with some abstract `ConsensusDecision`, it may be easier to implement nodes with different type of consensus (Casper, GRANDPA, or even plain simple PoW).
However, is this interesting to us, is a big question though.